app/src/main/java/org/thoughtcrime/securesms/database/helpers/SQLCipherMigrationHelper.java

```diff
@@ -0,0 +1,257 @@
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.annimon.stream.function.BiFunction;
+
+import net.zetetic.database.sqlcipher.SQLiteDatabase;
+
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.R;
+import org.thoughtcrime.securesms.crypto.AsymmetricMasterCipher;
+import org.thoughtcrime.securesms.crypto.AttachmentSecretProvider;
+import org.thoughtcrime.securesms.crypto.MasterCipher;
+import org.thoughtcrime.securesms.crypto.MasterSecret;
+import org.thoughtcrime.securesms.crypto.MasterSecretUtil;
+import org.thoughtcrime.securesms.migrations.LegacyMigrationJob;
+import org.thoughtcrime.securesms.service.GenericForegroundService;
+import org.thoughtcrime.securesms.util.Base64;
+import org.thoughtcrime.securesms.util.TextSecurePreferences;
+import org.whispersystems.libsignal.InvalidMessageException;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+public class SQLCipherMigrationHelper {
+
+  private static final String TAG = Log.tag(SQLCipherMigrationHelper.class);
+
+  private static final long ENCRYPTION_SYMMETRIC_BIT  = 0x80000000;
+  private static final long ENCRYPTION_ASYMMETRIC_BIT = 0x40000000;
+
+  static void migratePlaintext(@NonNull Context context,
+                               @NonNull android.database.sqlite.SQLiteDatabase legacyDb,
+                               @NonNull SQLiteDatabase modernDb)
+  {
+    modernDb.beginTransaction();
+    int foregroundId = GenericForegroundService.startForegroundTask(context, context.getString(R.string.SQLCipherMigrationHelper_migrating_signal_database)).getId();
+    try {
+      copyTable("identities", legacyDb, modernDb, null);
+      copyTable("push", legacyDb, modernDb, null);
+      copyTable("groups", legacyDb, modernDb, null);
+      copyTable("recipient_preferences", legacyDb, modernDb, null);
+      copyTable("group_receipts", legacyDb, modernDb, null);
+      modernDb.setTransactionSuccessful();
+    } finally {
+      modernDb.endTransaction();
+      GenericForegroundService.stopForegroundTask(context, foregroundId);
+    }
+  }
+
+  public static void migrateCiphertext(@NonNull Context context,
+                                       @NonNull MasterSecret masterSecret,
+                                       @NonNull android.database.sqlite.SQLiteDatabase legacyDb,
+                                       @NonNull SQLiteDatabase modernDb,
+                                       @Nullable LegacyMigrationJob.DatabaseUpgradeListener listener)
+  {
+    MasterCipher           legacyCipher           = new MasterCipher(masterSecret);
+    AsymmetricMasterCipher legacyAsymmetricCipher = new AsymmetricMasterCipher(MasterSecretUtil.getAsymmetricMasterSecret(context, masterSecret));
+
+    modernDb.beginTransaction();
+
+    int foregroundId = GenericForegroundService.startForegroundTask(context, context.getString(R.string.SQLCipherMigrationHelper_migrating_signal_database)).getId();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *startForegroundTask* method in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).